subsys/net/openthread/rpc/server/ot_rpc_srp_client.c

```diff
@@ -0,0 +1,812 @@
+
+#include <zephyr/net/openthread.h>
+#include <zephyr/sys/slist.h>
+
+#include <openthread/srp_client.h>
+#include <openthread/srp_client_buffers.h>
+
+#include <string.h>
+
+struct host_data {
+	char hostname[OT_DNS_MAX_NAME_SIZE];
+	otIp6Address *addresses;
+	size_t addresses_num;
+};
+
+struct service_data {
+	sys_snode_t node;
+	uintptr_t client_id;
+	otSrpClientService service;
+	uint8_t *buffer;
+};
+
+static struct host_data host_data;
+static sys_slist_t service_data = SYS_SLIST_STATIC_INIT(&service_data);
+
+static void host_data_clear(void)
+{
+	if (host_data.addresses) {
+		free(host_data.addresses);
+		host_data.addresses = NULL;
+	}
+
+	memset(&host_data, 0, sizeof(host_data));
+}
+
+static void service_data_clear(void)
+{
+	struct service_data *service;
+	struct service_data *next;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&service_data, service, next, node) {
+		if (service->buffer) {
+			free(service->buffer);
+		}
+
+		free(service);
+	}
+
+	sys_slist_init(&service_data);
+}
+
+static struct service_data *service_data_find(uintptr_t client_id, struct service_data **prev)
+{
+	struct service_data *service;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&service_data, service, node) {
+		if (service->client_id == client_id) {
+			return service;
+		}
+
+		if (prev) {
+			*prev = service;
+		}
+	}
+
+	return service;
+}
+
+static struct service_data *service_data_find_prev(struct service_data *service)
+{
+	sys_snode_t *prev;
+
+	if (sys_slist_find(&service_data, &service->node, &prev)) {
+		return CONTAINER_OF(prev, struct service_data, node);
+	}
+
+	return NULL;
+}
+
+static void service_buffers_free(struct service_data *service)
+{
+	if (service->buffer) {
+		free(service->buffer);
+	}
+
+	free(service);
+}
+
+static void service_data_free(struct service_data *service, struct service_data *prev)
+{
+	sys_slist_remove(&service_data, (prev ? &prev->node : NULL), &service->node);
+
+	service_buffers_free(service);
+}
+
+static size_t service_list_length(const otSrpClientService *services)
+{
+	size_t length = 0;
+
+	for (const otSrpClientService *s = services; s != NULL; s = s->mNext) {
+		++length;
+	}
+
+	return length;
+}
+
+static const char *decode_str(struct nrf_rpc_cbor_ctx *ctx, uint8_t **buffer_pos,
+			      size_t *buffer_size)
+{
+	/* Decode string and update buffer position and current size */
+	size_t str_size;
+	const char* start;
+
+	start = nrf_rpc_decode_str(ctx, *buffer_pos, *buffer_size);
+
+	str_size = strlen(*buffer_pos) + 1;
+	*buffer_size -= str_size;
+	*buffer_pos += str_size;
+
+	return start;
+}
+
+static void decode_subtypes(struct nrf_rpc_cbor_ctx *ctx, otSrpClientService *service,
+			    uint8_t **buffer_pos, size_t *buffer_size)
+{
+	const char **ptrs = (const char **)*buffer_pos;
+	size_t index = 0;
+
+	zcbor_list_start_decode(ctx->zs);
+
+	/* Prepare storage for array of pointers in the buffer */
+	*buffer_pos += sizeof(const char *) * (ctx->zs->elem_count + 1);
+
+	while (ctx->zs->elem_count) {
```

Comment:
Has this been tested? My understanding was that since we don't build our samples with `ZCBOR_CANONICAL`, zcbor will use indefinite length arrays, and then during decoding it will put a large number to `elem_count` :)
```
	if (!zcbor_new_backup(state, indefinite_length_array
				? ZCBOR_LARGE_ELEM_COUNT : new_elem_count)) {
		FAIL_RESTORE();
	}
```